### PR TITLE
feature: preserve body after parse

### LINKED ISSUE #41: How to reconstruct body after upload
I'm trying to make an anti-virus middle ware which first check the  uploaded files with help of lua-resty-upload and if they are not infected proxy_pass them to some dynamic proxy target.
I was able to handle the first part easily thanks to lua-resty-upload but the issue is when I proxy_pass the request after anti-virus check the body is empty. which seems pretty normal because lua-resty-upload has consumed the body already.
So I should reconstruct the body again after virus check, But I don't have the raw binary data of request at that point to easily reconstruct the body, all I have is parsed body which is hard to be converted to raw binary again.
Is there any shortcut method which can give me the raw binary data after process of lua-resty-upload is finished?

### PR DESCRIPTION
**Pull request by suikabreaker**
Fix #41 and #42. Inspired from this PR: #43. 

My goal is to leave the request body unchanged after "regeneration" and decouple the implementation of #61 from this, so I implement this in another way and try to make fewer modifications to the original logic.

### CODE REVIEW

**Comment by suikabreaker** [comment]
@zhuizhuhaomeng May you review this PR?

File: lib/resty/upload.lua [hunk 1]
```diff
@@ -21,12 +24,52 @@ local STATE_READING_HEADER = 2
 local STATE_READING_BODY = 3
 local STATE_EOF = 4
 
+local function warped_receiveuntil(self, until_str)
```
Comment (xiaocang) [inline lib/resty/upload.lua:27]: maybe the function name should be `wrapped_receiveuntil`

Comment (xiaocang) [reply lib/resty/upload.lua:27]: these two local functions (this one and the `req_socket_body_collector` one) are more suitable to be placed below all the local variables (line 72)

File: lib/resty/upload.lua [hunk 2]
```diff
@@ -21,12 +24,52 @@ local STATE_READING_HEADER = 2
...
+    if iter == nil then
+        ngx_finish_body()
+    end
+
+    local function warped(size)
+        local ret, err = iter(size)
+        if ret then
+            ngx_append_body(ret)
+        end
+        -- non-nil ret for call with no size or successful size call and nil ret
```
Comment (xiaocang) [inline lib/resty/upload.lua:38]: you can put one extra blank line here before to make the code easier to read
```suggestion

        -- non-nil ret for call with no size or successful size call and nil ret
```

File: lib/resty/upload.lua [hunk 3]
```diff
@@ -21,12 +24,52 @@ local STATE_READING_HEADER = 2
...
+        if (not size and ret) or (size and not ret and not err) then
+            ngx_append_body(until_str)
+        end
+        return ret, err
+    end
+    return warped, err_outer
```
Comment (xiaocang) [inline lib/resty/upload.lua:44]: Ditto
```suggestion

    return warped, err_outer
```

File: lib/resty/upload.lua [hunk 4]
```diff
@@ -21,12 +24,52 @@ local STATE_READING_HEADER = 2
...
+end
+
+local function warped_receive(self, arg)
+    local ret, err, partial = self:old_receive(arg)
+    if ret then
+        ngx_append_body(ret)
+    elseif partial then
```
Comment (xiaocang) [inline lib/resty/upload.lua:51]: Ditto
```suggestion

    elseif partial then
```

File: lib/resty/upload.lua [hunk 5]
```diff
@@ -21,12 +24,52 @@ local STATE_READING_HEADER = 2
...
+        ngx_append_body(partial)
+    end
+
+    if ret == nil then
+        ngx_finish_body()
+    end
+    return ret, err
```
Comment (xiaocang) [inline lib/resty/upload.lua:58]: Ditto
```suggestion

    return ret, err
```

File: lib/resty/upload.lua [hunk 6]
```diff
@@ -104,6 +153,9 @@ local function discard_line(self)
 
     local dummy, err = read_line(1)
     if dummy then
+        if self.preserve_body then
+            ngx_finish_body()
+        end
```
Comment (xiaocang) [inline lib/resty/upload.lua:158]: Ditto
```suggestion
        end

```

File: lib/resty/upload.lua [hunk 7]
```diff
@@ -179,6 +231,9 @@ local function read_header(self)
...
```
Comment (xiaocang) [inline lib/resty/upload.lua:236]: Ditto
```suggestion
        end
 
```

File: t/sanity.t [hunk 8]
```diff
@@ -648,3 +648,128 @@ read: ["eof"]
...
+\r\n-----------------------------820127721219505131303151179--\r
+}
+--- response_body
+content_length: 336
+remain body length: 336
+
+--- no_error_log
+[error]
+
+
```
Comment (xiaocang) [inline t/sanity.t:775]: There's no need to have extra blank lines at the end of the file.
```suggestion
```

File: t/sanity.t [hunk 9]
```diff
@@ -648,3 +648,128 @@ read: ["eof"]
...
+            form:set_timeout(1000) -- 1 sec
+
+            while true do
+                local typ, res, err = form:read()
+                if not typ then
+                    ngx.say("failed to read: ", err)
+                    return
+                end
+
+
```
Comment (xiaocang) [inline t/sanity.t:736]: One blank line is enough here

File: lib/resty/upload.lua [hunk 10]
```diff
@@ -21,12 +24,52 @@ local STATE_READING_HEADER = 2
...
+    elseif partial then
+        ngx_append_body(partial)
+    end
+
+    if ret == nil then
+        ngx_finish_body()
+    end
+    return ret, err
+end
+
```
Comment (xiaocang) [inline lib/resty/upload.lua:60]: Between two local functions, it is better to use two blank lines

**Review by xiaocang — CHANGES_REQUESTED** [review]

**Review by xiaocang — COMMENTED** [review]
The document of the `preserve` option is also needed.

Note that it should be stated that if the `preserve` option is turned on, the memory usage is doubled.

**Comment by xiaocang** [comment]
And there is a small doubt on my own: what is the scenario for the use of the function of preserving the request body here, and is it common enough?

File: lib/resty/upload.lua (hunk 1 above)
Comment (suikabreaker) [inline lib/resty/upload.lua:27]:
```suggestion
local function warpped_receiveuntil(self, until_str)
```

File: lib/resty/upload.lua (hunk 10 above)
Comment (suikabreaker) [inline lib/resty/upload.lua:60]:
```suggestion


```

File: t/sanity.t (hunk 9 above)
Comment (suikabreaker) [inline t/sanity.t:736]:
```suggestion
```

**Comment by suikabreaker** [comment]
> And there is a small doubt on my own: what is the scenario for the use of the function of preserving the request body here, and is it common enough?

There is. I have been working on a WAF project, which acts as a proxy filtering request body. It needs to pass the body(and if we can, do not modify the body) if there's no attempt of attack found.

File: lib/resty/upload.lua [hunk 11]
```diff
@@ -21,12 +24,57 @@ local STATE_READING_HEADER = 2
...
+    end
+
+    return ret, err
+end
+
+
+local function req_socket_body_collector(sock)
+    sock.old_receiveuntil = sock.receiveuntil
+    sock.old_receive = sock.receive
+    sock.receiveuntil = warped_receiveuntil
```
Comment (suikabreaker) [inline lib/resty/upload.lua:69]:
```suggestion
    sock.receiveuntil = warpped_receiveuntil
```

**Comment by suikabreaker** [comment]
> The document of the `preserve` option is also needed.
> 
> Note that it should be stated that if the `preserve` option is turned on, the memory usage is doubled.

Seems other two options are also not documented. Added description.

**Comment by xiaocang** [comment]
@suikabreaker there's still typo: warpped -> wrapped.

File: lib/resty/upload.lua [hunk 12]
```diff
@@ -21,11 +24,57 @@ local STATE_READING_HEADER = 2
...
+    end
+
+    return ret, err
+end
+
+
+local function req_socket_body_collector(sock)
+    sock.old_receiveuntil = sock.receiveuntil
+    sock.old_receive = sock.receive
+    sock.receiveuntil = warpped_receiveuntil
```
Comment (suikabreaker) [inline lib/resty/upload.lua:74]: warpped
```suggestion
    sock.receiveuntil = wrapped_receiveuntil
```

File: lib/resty/upload.lua [hunk 13]
```diff
@@ -21,11 +24,57 @@ local STATE_READING_HEADER = 2
 local STATE_READING_BODY = 3
 local STATE_EOF = 4
 
-
 local mt = { __index = _M }
 
 local state_handlers
 
+local function warpped_receiveuntil(self, until_str)
```
Comment (suikabreaker) [inline lib/resty/upload.lua:31]: warpped
```suggestion
local function wrapped_receiveuntil(self, until_str)
```

**Comment by suikabreaker** [comment]
> @suikabreaker there's still typo: warpped -> wrapped.

Fixed.

File: lib/resty/upload.lua [hunk 14]
```diff
@@ -21,11 +24,57 @@ local STATE_READING_HEADER = 2
...
 
 local state_handlers
 
+local function wrapped_receiveuntil(self, until_str)
+    local iter, err_outer = self:old_receiveuntil(until_str)
+    if iter == nil then
+        ngx_finish_body()
+    end
+
+    local function warped(size)
```
Comment (xiaocang) [inline lib/resty/upload.lua:37]: there are still typos here and elsewhere below.

**Review by xiaocang — CHANGES_REQUESTED** [review]

**Comment by xiaocang** [comment]
The other part looks good to me.

File: lib/resty/upload.lua (hunk 14 above)
Comment (suikabreaker) [inline lib/resty/upload.lua:37]:
```suggestion
    local function wrapped(size)
```

File: lib/resty/upload.lua [hunk 15]
```diff
@@ -21,11 +24,57 @@ local STATE_READING_HEADER = 2
...
+        end
+
+        -- non-nil ret for call with no size or successful size call and nil ret
+        if (not size and ret) or (size and not ret and not err) then
+            ngx_append_body(until_str)
+        end
+        return ret, err
+    end
+
+    return warped, err_outer
```
Comment (suikabreaker) [inline lib/resty/upload.lua:50]:
```suggestion
    return wrapped, err_outer
```

File: lib/resty/upload.lua [hunk 16]
```diff
@@ -21,11 +24,57 @@ local STATE_READING_HEADER = 2
...
+end
+
+
+local function warped_receive(self, arg)
```
Comment (suikabreaker) [inline lib/resty/upload.lua:54]:
```suggestion
local function wrapped_receive(self, arg)
```

File: lib/resty/upload.lua [hunk 17]
```diff
@@ -21,11 +24,57 @@ local STATE_READING_HEADER = 2
...
+
+    return ret, err
+end
+
+
+local function req_socket_body_collector(sock)
+    sock.old_receiveuntil = sock.receiveuntil
+    sock.old_receive = sock.receive
+    sock.receiveuntil = wrapped_receiveuntil
+    sock.receive = warped_receive
```
Comment (suikabreaker) [inline lib/resty/upload.lua:75]:
```suggestion
    sock.receive = wrapped_receive
```

File: lib/resty/upload.lua (hunk 14 above)
Comment (suikabreaker) [reply lib/resty/upload.lua]: > there are still typos here and elsewhere below.

Searched more carefully. It should now be fixed.

**Comment by suikabreaker** [comment]
> The other part looks good to me.

I have re-requested review.

**Review by xiaocang — APPROVED** [review]

**Comment by xiaocang** [comment]
could you help check this PR?
ping @zhuizhuhaomeng @doujiang24

File: README.markdown [hunk 18]
```diff
@@ -161,6 +161,19 @@ not want to save the data on local file systems.
...
 [Back to TOC](#table-of-contents)
 
+Usage
+=====
+
+```lua
+local upload = require "resty.upload"
+local form, err = upload:new(self, chunk_size, max_line_size, preserve_body)
+```
+`chunk_size` defaults to 4096. It will be the size used to read the socket.
```
Comment (zhuizhuhaomeng) [inline README.markdown:171]:
```suggestion
`chunk_size` defaults to 4096. It is the size used to read data from the socket.
```

**Pull request merged**